pkg/client/amqp.go

```diff
@@ -0,0 +1,240 @@
+	}
+}
+
+//------------------------------------------------------------------------------
+
+// AMQPManager is a struct to manage AMQP connections
+type AMQPManagerConfig struct {
+	address  string
+	user     string
+	password string
+	config   *amqp.Config
+}
+
+// NewAMQPClient is AMQP client constructor
+func NewAMQPManagerConfig(address, user, password string, config *amqp.Config) *AMQPManagerConfig {
+	return &AMQPManagerConfig{
+		address:  address,
+		user:     user,
+		password: password,
+		config:   config,
+	}
+}
+
+// AMQPManager is a struct to manage AMQP connections
+type AMQPManager struct {
+	config     *AMQPManagerConfig
+	connection *amqp.Connection
+	pool       *Pool
+	shutdown   bool
+	Reconnects uint64
+	logger     *logrus.Logger
+	sync.Mutex
+}
+
+// NewAMQPClient is AMQP client constructor
+func (am *AMQPManager) newAMQPConnection() (err error) {
+	var connection *amqp.Connection
+	url := fmt.Sprintf(
+		"%s://%s:%s@%s/",
+		"amqp",
+		am.config.user,
+		am.config.password,
+		am.config.address,
+	)
+	if am.config.config != nil {
+		connection, err = amqp.DialConfig(
+			url,
+			*am.config.config,
+		)
+	} else {
+		connection, err = amqp.Dial(url)
+	}
+	if err != nil {
+		return helper.ErrorPrintf("unable to start AMQP subsytem: %v", err)
+	}
+	am.Lock()
+	am.connection = connection
+	am.Unlock()
+	return nil
+}
+
+// NewAMQPManager is AMQPManager constructor
+func NewAMQPManager(config *AMQPManagerConfig, maxCap uint, logger *logrus.Logger) (*AMQPManager, error) {
+	manager := &AMQPManager{
+		config: config,
+		pool:   NewPool(maxCap, nil, logger),
+		logger: logger,
+	}
+	err := manager.newAMQPConnection()
+	if err != nil {
+		return nil, err
+	}
+	go manager.reconnectConn()
+	return manager, nil
+}
+
+// SetPoolFactory adds func factory to pool
+func (am *AMQPManager) SetPoolFactory(factory Factory) {
+	am.pool.SetFactory(factory)
+}
+
+// AddClient adds a new Channel in pool
+func (am *AMQPManager) AddClient() error {
+	channel, err := am.connection.Channel()
+	if err != nil {
+		return err
+	}
+	wrapper := NewAMQPWrapper(channel)
+	go am.reconnectChannel(wrapper)
+	am.pool.Put(wrapper)
+	return nil
+}
+
+// GetChannel get a new AMQP connection channel in pool
+func (am *AMQPManager) GetClient() (*AMQPWrapper, error) {
+	amqpWrapper := am.pool.Get()
+	if amqpWrapper == nil {
+		return nil, helper.ErrorPrint("no AMQP client available")
+	}
+	return amqpWrapper.(*AMQPWrapper), nil
+}
+
+// PutClient add an existing AMQP client in pool
+func (am *AMQPManager) PutClient(client *AMQPWrapper) {
+	err := am.pool.Put(client)
+	if err != nil {
+		am.logger.Warn(err)
+		err = client.Channel.Close()
+		if err != nil {
+			am.logger.WithError(err).Warn("An error occurred while closing channel")
+		}
+	}
+}
+
+// Close closes AMQP connection and channels
+func (am *AMQPManager) Close() error {
+	am.shutdown = true
+	err := am.connection.Close()
+	if err != nil {
+		am.shutdown = false
+		return err
+	}
+	return nil
```

Comment:
Faut que je trouve un moyen de mieux handle le `shutdown`. Cela fonctionne en l'état mais ce n'est pas optimal.

Reply:
Ensure the connection is closed in loops to handle edge cases.